File: 15.3-sum.cpp
/*
 * @lc app=leetcode id=15 lang=cpp
 *
 * [15] 3Sum
 */
#include <bits/stdc++.h>

using namespace std;

// @lc code=start
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
        sort(nums.begin(), nums.end());
        vector<vector<int>> result;

        // Two Pointers
        for (int i=0; i<n; i++) {
            if (nums[i] > 0) break;
            if (i>0 && nums[i]==nums[i - 1]) { continue; }

            int L = i + 1;
            int R = n - 1;
            while (L < R) {
                int sum = nums[i] + nums[L] + nums[R];
                if (sum < 0) { L++; }
                if (sum > 0) { R--; }
                if (sum == 0) {
                    result.push_back(vector<int>{nums[i], nums[L], nums[R]});
                    L++;
                    R--;
                    while (L<R && nums[L]==nums[L-1]) { L++; }
                }
            }
        }
        
        return result;
    }
};
// @lc code=end

File: 15.3-sum.py
#
# @lc app=leetcode id=15 lang=python3
#
# [15] 3Sum
#
from typing import List, Set, Tuple

# @lc code=start
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res: Set[Tuple[int]] = set()
        nums.sort()
        n = len(nums)
        for j in range(1, n-1):
            i = j-1
            k = j+1
            while i >= 0 and k < n:
                s = nums[i] + nums[j] + nums[k]
                if s > 0:
                    i -= 1
                elif s < 0:
                    k += 1
                else:
                    res.add((nums[i], nums[j], nums[k]))
                    k += 1
        return list(res)
# @lc code=end

def main():
    s = Solution()
    print(s.threeSum([0, 0, 0]))

if __name__ == '__main__':
    main()